Remap active issue urls to use runtime instead of corefx/coreclr

safern opened the pull request:
Used an automated console app using GH Rest API to get the new issue URL:

```cs
static async Task Main(string[] args)
        {
            string[] files = Directory.GetFiles(@"E:\repos\runtime\src\libraries", "*.cs", SearchOption.AllDirectories);
            Regex regex = new Regex("((http(s)?(\\:\\/\\/))+(www\\.)?([\\w\\-\\.\\/])*(\\.[a-zA-Z]{2,3}\\/dotnet?))[^.,;:\\?\\!\\@\\^\\$ -][^\\s\\b\\n \\)\".,;:]*");
            int count = 0;
            int requestNum = 0;
            Dictionary<string, string> issueMapping = new Dictionary<string, string>();
            var github = new GitHubClient(new ProductHeaderValue("IssueMapper"))
            {
                Credentials = new Credentials("user", "PAT")
            };

            foreach (var file in files)
            {
                string text = File.ReadAllText(file, new UTF8Encoding(true));
                var matches = regex.Matches(text);
                List<string> issuesFound = new List<string>();
                foreach (Match match in matches)
                {
                    if (match.Value.Contains("corefx") || match.Value.Contains("coreclr"))
                    {
                        count++;
                        string issueUrl = match.Value;
                        if (issueMapping.ContainsKey(issueUrl))
                        {
                            if (!issuesFound.Contains(issueUrl))
                                issuesFound.Add(issueUrl);
                            continue;
                        }
                        if (!Uri.IsWellFormedUriString(issueUrl, UriKind.Absolute) || !issueUrl.Contains("issues/"))
                            continue;

                        Uri uri = new Uri(issueUrl);
                        string[] segments = uri.Segments;
                        int issueNumber = int.Parse(segments[4].TrimEnd('/'));
                        string repo = segments[2].TrimEnd('/');
                        requestNum++;

                        if (issueNumber <= 0)
                            continue;

                        Issue issue = await github.Issue.Get("dotnet", repo, issueNumber);
                        string newUrl = issue.HtmlUrl;
                        if (!string.IsNullOrEmpty(uri.Fragment))
                            newUrl += uri.Fragment;

                        issuesFound.Add(issueUrl);
                        issueMapping.Add(issueUrl, newUrl);

                        Thread.Sleep(100);

                    }
                }

                bool textChanged = false;
                foreach (string issue in issuesFound)
                {
                    textChanged = true;
                    text = text.Replace(issue, issueMapping[issue]);
                }

                if (textChanged)
                {
                    Encoding enc = GetEncoding(file);
                    File.WriteAllText(file, text, enc);
                }
            }


            Console.WriteLine($"Total Count: {count}");
            Console.WriteLine($"Total Requests: {requestNum}");
            Console.WriteLine($"FIles Searched: {files.Length}");
        }

        static Encoding GetEncoding(string filename)
        {
            byte[] bom = new byte[4];
            using (FileStream file = File.OpenRead(filename))
            {
                file.Read(bom, 0, 4);
            }

            if (bom[0] == 0x2b && bom[1] == 0x2f && bom[2] == 0x76) return Encoding.UTF7;
            if (bom[0] == 0xef && bom[1] == 0xbb && bom[2] == 0xbf) return Encoding.UTF8;
            if (bom[0] == 0xff && bom[1] == 0xfe) return Encoding.Unicode; //UTF-16LE
            if (bom[0] == 0xfe && bom[1] == 0xff) return Encoding.BigEndianUnicode; //UTF-16BE
            if (bom[0] == 0 && bom[1] == 0 && bom[2] == 0xfe && bom[3] == 0xff) return Encoding.UTF32;
            return Encoding.ASCII;
        }
```

cc: @jkotas @stephentoub @ViktorHofer @jaredpar @danmosemsft

safern commented on src/libraries/Common/tests/Tests/System/StringTests.cs:
```diff
@@ -7078,7 +7078,7 @@ public static void StartsWithNoMatchNonOrdinal_StringComparison()
         // NOTE: This is by design. Unix ignores the null characters (i.e. null characters have no weights for the string comparison).
         // For desired behavior, use ordinal comparison instead of linguistic comparison.
-        // This is a known difference between Windows and Unix (https://github.com/dotnet/coreclr/issues/2051).
+        // This is a known difference between Windows and Unix (https://github.com/dotnet/corefx/pull/2051).
```
These ones was an accident on a search and replace I was doing.

safern commented:
This helps re-enabling tests and finding disabled tests when an issue is investigated, because now when looking at it we get the runtime URL.

ahsonkhan commented:
Do you plans to fix-up links found in source/comments outside of `[ActiveIssue]`?

For example, see these in S.T.J:
https://github.com/dotnet/runtime/blob/82389e0e403d6e66052a4950b963e188af26504c/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs#L383
https://github.com/dotnet/runtime/blob/82389e0e403d6e66052a4950b963e188af26504c/src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs#L608
https://github.com/dotnet/runtime/blob/82389e0e403d6e66052a4950b963e188af26504c/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs#L70

ahsonkhan reviewed (commented):
S.T.J and S.M changes LGTM.

Skimmed through the rest and they seem fine.

safern commented:
I just noticed a small bug in my console app where the base url is always Corefx so the mapping of coreclr issues might be wrong mapped to Corefx pulls.

Will fix and update.

safern commented:
Yeah I’ll do a pass with a regex that doesn’t include ActiveIssue to update all URLS.

jkotas commented:
> some of our test files weren't encoded in UTF8

We assume UTF8 encoding for all source files. And majority of the files in the repo are without the explicit BOM because of it just looks ugly in number of Unix tools.

The checked in copy of src/libraries/Common/tests/System/Net/Http/DefaultCredentialsTest.cs does not have BOM, but it looks like you are adding the BOM to this file. I do not think it is an improvement. You should save the files without the BOM in your script.

safern committed:
Update all issues in source code to point to runtime

safern committed:
Merge remote-tracking branch 'dotnet/master' into RemapActiveIssueToRuntime

safern commented:
>  I do not think it is an improvement. You should save the files without the BOM in your script.

Ok, thanks. I updated the script to preserve the Encoding based on the BOM it contained and no BOM changes are present.

Also, I fixed a few bugs in my script and now doing a search, I couldn't find references to corefx/coreclr issues in the code, except for pull requests, but since those weren't transferred we can leave them as is.

ViktorHofer commented:
Do all ActiveIssue ids map to an issue? Wondering if devs also used pr ids here which would be wrong.

safern commented:
> Do all ActiveIssue ids map to an issue? Wondering if devs also used pr ids here which would be wrong.

I’m not sure. I will do an audit later after this is merged with a variation of my script. Since this PR is big enough I didn’t want to include other than remapping of issues.

safern committed:
Merge remote-tracking branch 'dotnet/master' into RemapActiveIssueToRuntime

safern commented:
Can someone please take a look when free? This is prone to merge conflicts, so it would be nice to merge soon.

safern committed:
Merge remote-tracking branch 'dotnet/master' into RemapActiveIssueToRuntime

safern commented:
Thanks @jkotas -- build was green before last small merge conflicts. Will go ahead and merge now.

ahsonkhan commented on src/libraries/System.Text.Json/tests/Serialization/Array.ReadTests.cs:
```diff
@@ -96,7 +96,7 @@ public static void ReadByteListAsJsonArray()
         public static void DeserializeObjectArray_36167()
         {
-            // https://github.com/dotnet/corefx/issues/36167
+            // https://github.com/dotnet/runtime/issues/29019
```
With issue number changed, the test names should be updated (it stills says `_36167`). We should just remove the issue number from it to avoid having to keep updating it. Same elsewhere (like `DictionaryOfObject_37569`).

safern replied:
Thanks @ahsonkhan... I will follow up on a PR.

ahsonkhan commented on src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.TryGet.Date.cs:
```diff
@@ -149,7 +149,7 @@ public static void TestingStringsInvalidConversionToDateTimeOffset(string jsonSt
         [Fact]
-        // https://github.com/dotnet/corefx/issues/39067.
+        // https://github.com/dotnet/runtime/issues/30095.
         public static void Regression39067_TestingDateTimeMinValue()
```
Same issue here (test name with issue number no longer matches, we should remove the number).
